### PR TITLE
coupled sensorring and head in one instance

### PR DESCRIPTION
**Pull request by mathias-luedtke**
merged sensorring config into head config, added remaps and relays.
needs https://github.com/ipa320/cob_command_tools/pull/164

### CODE REVIEW

**Comment by NadiaHG** [comment]
@ipa-mdl tested on the robot?

**Comment by mathias-luedtke** [comment]
The spawner does not work very well if spawning fails for some reason.
Perhaps we should implements a service relay with controller spawning..

**Comment by mathias-luedtke** [comment]
> tested on the robot

I was able to move the sensorring ;)

**Comment by mathias-luedtke** [comment]
Perhaps we should substitute with https://github.com/ipa320/cob_control/pull/118

File: cob_bringup/package.xml [hunk 1]
```diff
@@ -93,6 +93,7 @@
   <exec_depend>schunk_powercube_chain</exec_depend>
   <exec_depend>schunk_sdh</exec_depend>
   <exec_depend>sick_visionary_t_driver</exec_depend>
+  <exec_depend>service_tools</exec_depend>
```
Comment (fmessmer) [inline cob_bringup/package.xml:95]: pedantic: keep alphabetic order :wink:

**Review by fmessmer — CHANGES_REQUESTED** [review]

File: cob_bringup/robots/cob4-10.xml [hunk 2]
```diff
@@ -124,15 +124,20 @@
...
+	          <node name="sensorring_load_relay" pkg="service_tools" type="service_relay" args="controller_manager/load_controller /head/controller_manager/load_controller" respawn="false" output="screen"/>
+	          <node name="sensorring_switch_relay" pkg="service_tools" type="service_relay" args="controller_manager/switch_controller /head/controller_manager/switcj_controller" respawn="false" output="screen"/>
+	          <node name="sensorring_joint_states_relay" pkg="topic_tools" type="relay" args="/head/joint_states joint_states" respawn="false" output="screen"/>
+
+	          <rosparam param="joint_names">[sensorring_joint]</rosparam>
+	          <param name="joint_trajectory_controller_name" value="sensorring_joint_trajectory_controller">
+	          <param name="sensorring_joint_trajectory_controller" value="forward">
+	          <node pkg="cob_control_mode_adapter" type="cob_control_mode_adapter_node" name="control_mode_adapter" cwd="node" respawn="false" output="screen"/>
+        </group>
+        <remap from="/head/sensorring_joint_trajectory_controller/follow_joint_trajectory" to="/sensorring/joint_trajectory_controller/follow_joint_trajectory"/>
```
Comment (fmessmer) [inline cob_bringup/robots/cob4-10.xml:140]: why is this magic needed?

is it true, that in the end, there will be no topics/actions/services under the ```sensorring``` namespace?

File: cob_hardware_config/cob4-10/config/joint_state_publisher.yaml [hunk 3]
```diff
@@ -1,2 +1,2 @@
 rate: 100 #Hz
-source_list: [/base/joint_states, /torso/joint_states, /sensorring/joint_states, /head/joint_states]
+source_list: [/base/joint_states, /torso/joint_states, /head/joint_states]
```
Comment (fmessmer) [inline cob_hardware_config/cob4-10/config/joint_state_publisher.yaml:2]: if you keep the ```sensorring/joint_states``` here, you would not need the topic_relay, right?

File: cob_bringup/robots/cob4-10.xml [hunk 4]
```diff
@@ -124,15 +124,20 @@
...
-        <include file="$(find cob_bringup)/controllers/generic_controller.launch">
-            <arg name="robot" value="$(arg robot)"/>
-            <arg name="component_name" value="sensorring"/>
-        </include>
+
+        <group name="sensorring">
+	          <node name="sensorring_init_relay" pkg="service_tools" type="service_relay" args="driver/init /head/driver/init" respawn="false" output="screen"/>
+	          <node name="sensorring_recover_relay" pkg="service_tools" type="service_relay" args="driver/recover /head/driver/recover" respawn="false" output="screen"/>
+	          <node name="sensorring_load_relay" pkg="service_tools" type="service_relay" args="controller_manager/load_controller /head/controller_manager/load_controller" respawn="false" output="screen"/>
+	          <node name="sensorring_switch_relay" pkg="service_tools" type="service_relay" args="controller_manager/switch_controller /head/controller_manager/switcj_controller" respawn="false" output="screen"/>
```
Comment (fmessmer) [inline cob_bringup/robots/cob4-10.xml]: typo: switcj -> switch

**Review by fmessmer — CHANGES_REQUESTED** [review]

File: cob_bringup/package.xml (hunk 1 above)
Comment (mathias-luedtke) [reply cob_bringup/package.xml:95]: Again? ;)
At least I hit the S range :D

File: cob_bringup/robots/cob4-10.xml (hunk 2 above)
Comment (mathias-luedtke) [reply cob_bringup/robots/cob4-10.xml:140]: This relocates the sensorring trajectory controller (spawned in head driver) into the sensorring namespace

File: cob_hardware_config/cob4-10/config/joint_state_publisher.yaml (hunk 3 above)
Comment (mathias-luedtke) [reply cob_hardware_config/cob4-10/config/joint_state_publisher.yaml:2]: `/head/joint_states` contains head+sensorring.
The relay is needed just for the script server, which expects them in sensorring as well.

File: cob_bringup/robots/cob4-10.xml (hunk 2 above)
Comment (mathias-luedtke) [reply cob_bringup/robots/cob4-10.xml:140]: The sensorring will almost look like all other component namespaces..

File: cob_bringup/package.xml [hunk 5]
```diff
@@ -92,8 +92,8 @@
   <exec_depend>rviz</exec_depend>
   <exec_depend>schunk_powercube_chain</exec_depend>
   <exec_depend>schunk_sdh</exec_depend>
-  <exec_depend>sick_visionary_t_driver</exec_depend>
```
Comment (mathias-luedtke) [inline cob_bringup/package.xml:95]: this went wrong, will push it into the first commit

**Comment by mathias-luedtke** [comment]
Fixed the typo and sorting.

Please note: I have not tested the last commit on the robot.

**Review by fmessmer — APPROVED** [review]

**Comment by fmessmer** [comment]
Will we need to apply these changes to cob4-7, too?
Let's also wait for travis and a "Tested on robot: OK"

**Comment by NadiaHG** [comment]
Still not working on the robot something with the namespaces is  wrong, after init I have the topics:

/head/joint_group_position_controller/command
/head/joint_group_velocity_controller/command
/head/joint_states
/head/joint_trajectory_controller/command
/head/joint_trajectory_controller/follow_joint_trajectory/cancel
/head/joint_trajectory_controller/follow_joint_trajectory/feedback
/head/joint_trajectory_controller/follow_joint_trajectory/goal
/head/joint_trajectory_controller/follow_joint_trajectory/result
/head/joint_trajectory_controller/follow_joint_trajectory/status
/head/joint_trajectory_controller/state
**/head/sensorring_joint_trajectory_controller/command
/head/sensorring_joint_trajectory_controller/state**

And 

/sensorring/joint_states
/sensorring/joint_trajectory_controller/follow_joint_trajectory/cancel
/sensorring/joint_trajectory_controller/follow_joint_trajectory/feedback
/sensorring/joint_trajectory_controller/follow_joint_trajectory/goal
/sensorring/joint_trajectory_controller/follow_joint_trajectory/result
/sensorring/joint_trajectory_controller/follow_joint_trajectory/status

and `rosservice call /head/controller_manager/list_controllers` says:

```
controller: 
  - 
    name: joint_trajectory_controller
    state: running
    type: position_controllers/JointTrajectoryController
    hardware_interface: hardware_interface::PositionJointInterface
    resources: ['head_2_joint', 'head_3_joint']
  - 
    name: joint_state_controller
    state: running
    type: joint_state_controller/JointStateController
    hardware_interface: hardware_interface::JointStateInterface
    resources: []
  - 
    name: sensorring_joint_trajectory_controller
    state: running
    type: position_controllers/JointTrajectoryController
    hardware_interface: hardware_interface::PositionJointInterface
    resources: ['sensorring_joint']
  - 
    name: joint_group_position_controller
    state: stopped
    type: position_controllers/JointGroupPositionController
    hardware_interface: hardware_interface::PositionJointInterface
    resources: ['head_2_joint', 'head_3_joint']
  - 
    name: joint_group_velocity_controller
    state: stopped
    type: velocity_controllers/JointGroupVelocityController
    hardware_interface: hardware_interface::VelocityJointInterface
    resources: ['head_2_joint', 'head_3_joint']
```

If I send a trajectory I see that I sending a command to /head/joint_trajectory_controller/command but the actuator doesn't move.

Tested with https://github.com/ipa320/cob_command_tools/pull/164 and https://github.com/ipa320/cob_control/pull/118.

@ipa-fxm I would like to wait until it is working before merge the config for the other robots, actually cob4-1, cob4-2, cob4-7, cob4-10 and soon cob4-5 should use this head/sensorring configuration.

**Comment by mathias-luedtke** [comment]
The topics are fine!
`/head/sensorring_joint_trajectory_controller/command` is used by rqt only and it expects it in the controller's namespace.
Not sure about `/head/sensorring_joint_trajectory_controller/state`, we don't use it anyway (?).

Head does not move right now since the motors cannot be started( check diagnostics, @ipa-tif needs to fix this).

**Comment by mathias-luedtke** [comment]
Before merging it for the other robots, the head+sensorring combo should be factored out into a separate launch file.
It would be better to actually split the controller config again and upload both.

**Comment by mathias-luedtke** [comment]
where to put such a launch file?

**Comment by mathias-luedtke** [comment]
rebased to keep the sensorring_controller.yaml (used with gazebo)

**Comment by fmessmer** [comment]
> where to put such a launch file?

As it only launches controller-related things, I'd put it into cob_bringup/controllers.

**Comment by mathias-luedtke** [comment]
> As it only launches controller-related things

It launches as driver and set-ups the controllers, i.e. `head_sensorring.xml`.

**Pull request closed without merging**